### PR TITLE
Split validation & serialization

### PR DESCRIPTION
**Pull request by ppershing**

### CODE REVIEW

File: src/interactions/signTx.ts [hunk 1]
```diff
@@ -9,10 +9,18 @@ import type {
   Withdrawal,
 } from "../Ada";
 import {
-  CertificateType,
   Errors,
 } from "../Ada"
-import cardano, { ParsedCertificate, SignTxIncluded } from "../cardano";
+import cardano, { SignTxIncluded } from "../cardano";
+import type {
```
Comment (refi93) [inline src/interactions/signTx.ts:15]: can't those three import statements be merged into a single one as they are apparently importing from the same location?

File: src/parsing.ts [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+    SingleHostIPRelay,
+    SingleHostNameRelay,
+    TxOutputTypeAddress,
+    TxOutputTypeAddressParams,
+    Withdrawal,
+} from "./Ada";
+// TODO: remove this dependency
+import { serializeOutputBasicParams } from "./cardano";
+import { TxErrors } from "./txErrors";
+import utils, { MAX_LOVELACE_SUPPLY_STR, Precondition } from "./utils";
```
Comment (refi93) [inline src/parsing.ts:20]: those two imports can also be merged, maybe worth enabling no-duplicate-imports rule https://eslint.org/docs/rules/no-duplicate-imports

File: src/cardano.ts [hunk 3]
```diff
@@ -806,38 +304,24 @@ export function serializePoolMetadata(
   }
 }
 
-export function serializeGetExtendedPublicKeyParams(path: BIP32Path): Buffer {
-  Precondition.check(!!path, GetKeyErrors.INVALID_PATH);
-
-  return utils.path_to_buf(path);
+export function serializeGetExtendedPublicKeyParams(path: ValidBIP32Path): Buffer {
+  return Buffer.concat([
+    utils.path_to_buf(path),
```
Comment (refi93) [inline src/cardano.ts:309]: why concatenating a single buffer? Isn't it equivalent to returning `utils.path_to_buf(path)` directly?

File: src/parsing.ts [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+        Precondition.checkIsPositiveUint64Str(
+            validityIntervalStartStr,
+            TxErrors.VALIDITY_INTERVAL_START_INVALID
+        );
+    }
+}
+
+
+
+export type VarlenAsciiString = string & { __type: 'ascii' }
```
Comment (refi93) [inline src/parsing.ts:238]: nitpick - I would prefer to keep types visually separated from the functions - e.g. types at the top and functions at the bottom, especially types such as FixlenHexString or ValidBIP32Path look generic enough to not be randomly introduced in the middle of the file

File: src/parsing.ts [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+        default:
+            throw new Error(TxErrors.CERTIFICATE_INVALID);
+    }
+}
+
+export function parseCertificates(certificates: Array<Certificate>): Array<ParsedCertificate> {
+    Precondition.checkIsArray(certificates, TxErrors.CERTIFICATES_NOT_ARRAY);
+
+    const parsed = certificates.map(cert => parseCertificate(cert))
+
```
Comment (refi93) [inline src/parsing.ts:114]: redundant empty line

File: src/cardano.ts (hunk 3 above)
Comment (refi93) [reply src/cardano.ts:309]: if the function is that simple and it's used one, I would consider putting the `utils.path_to_buf(path)` call right where the function is invoked

File: src/parsing.ts [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+    PoolMetadataParams,
+    PoolOwnerParams,
+    PoolParams,
+    RelayParams,
+    SingleHostIPRelay,
+    SingleHostNameRelay,
+    TxOutputTypeAddress,
+    TxOutputTypeAddressParams,
+    Withdrawal,
+} from "./Ada";
```
Comment (refi93) [inline src/parsing.ts:16]: this makes for a cyclic import as Ada.ts imports the AddressTypeNibble and CertificateType from here - wouldn't it make sense to keep those types within Ada.ts as they are relevant for the library users as well?

File: src/parsing.ts [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+}
+
+export enum CertificateType {
+    STAKE_REGISTRATION = 0,
+    STAKE_DEREGISTRATION = 1,
+    STAKE_DELEGATION = 2,
+    STAKE_POOL_REGISTRATION = 3,
+}
+
+const KEY_HASH_LENGTH = 28;
```
Comment (refi93) [inline src/parsing.ts:38]: are we ok with keeping the duplicate of this constant in cardano.ts? is this just some temporary stage towards further reorganization of types/constants?

File: src/cardano.ts [hunk 8]
```diff
@@ -47,206 +21,6 @@ export const SignTxIncluded = Object.freeze({
 const KEY_HASH_LENGTH = 28;
```
Comment (refi93) [inline src/cardano.ts:21]: not sure if left out by mistake or on purpose but duplicates of KEY_HASH_LENGTH and TX_HASH_LENGTH are declared in parsing.ts

**Review by refi93 — CHANGES_REQUESTED** [review]
see comments - not really significant remarks, I'm mostly concerned about apparently duplicate constants and potentially better places for some types

File: src/parsing.ts (hunk 4 above)
Comment (refi93) [reply src/parsing.ts:238]: ok, not really, probably it's better like it is now, in the end, the types give additional insight when reading code that uses them

Comment (ppershing) [reply src/parsing.ts:238]: It is likely that the basic types + parsing functions eventually move to helpers anyway

File: src/parsing.ts (hunk 6 above)
Comment (ppershing) [reply src/parsing.ts:16]: Don't get me even start on the import cycles. I know. But that is for another type. We need quite a few restructurings, especially splitting external & internal types, etc.

File: src/interactions/signTx.ts (hunk 1 above)
Comment (ppershing) [reply src/interactions/signTx.ts:15]: Type imports cannot be mixed with normal imports. But yes, the second two can be merged

**Comment by ppershing** [comment]
Fixed most of the things

**Pull request merged**